packages/streamid/src/stream-ref-parsing.ts

```diff
@@ -0,0 +1,120 @@
+import { readCid, readVarint } from './reading-bytes.js'
+import { STREAMID_CODEC } from './constants.js'
+import { CID } from 'multiformats/cid'
+import { base36 } from 'multiformats/bases/base36'
+
+export type StreamIDComponents = {
+  kind: 'stream-id'
+  type: number
+  genesis: CID
+}
+
+export type CommitIDComponents = {
+  kind: 'commit-id'
+  type: number
+  genesis: CID
+  commit: CID | null
+}
+
+export type StreamRefComponents = StreamIDComponents | CommitIDComponents
+
+/**
+ * Parse StreamID or CommitID from bytes.
+ */
+export function fromBytes(input: Uint8Array, title = 'StreamRef'): StreamRefComponents {
+  const [streamCodec, streamCodecRemainder] = readVarint(input)
+  if (streamCodec !== STREAMID_CODEC)
+    throw new Error(`Invalid ${title}, does not include streamid codec`)
+  const [type, streamtypeRemainder] = readVarint(streamCodecRemainder)
+  const cidResult = readCid(streamtypeRemainder)
+  const [genesis, gnesisRemainder] = cidResult
+  if (gnesisRemainder.length === 0) {
+    return {
+      kind: 'stream-id',
+      type: type,
+      genesis: genesis,
+    }
+  } else if (gnesisRemainder.length === 1) {
+    // Zero commit
+    return {
+      kind: 'commit-id',
+      type: type,
+      genesis: genesis,
+      commit: null,
```

Comment:
this is technically a small behavior change as now you're not actually extracting the final byte, you're just assuming it's '0', which is the only value that is supported.  I'm not sure if it really matters, but if we wanted to preserve existing behavior I think it'd be as simple as allowing the `commit` field here to have type 'string' as well and return `genesisRemainder[0]` for the commit here.

Reply:
I do not understand the suggestion. This function parses bytes. Byte-encoded CommitID can only contain `[0]` as `genesisRemainder` or a full `CID`. What is the intent behind the suggestion?

Reply:
Your most recent commit made the behavior change even smaller, since now you check that `genesisRemainder[0] === 0` above.  I think now the only difference is if the final byte was something other than '0', in the old version we would throw the error `'Cannot specify commit as a number except to request commit 0 (the genesis commit)'`, whereas now we'll pass the extra byte into `readCid` and presumably get some error message out of that that it isn't a valid CID.

In the old version the genesisRemainder final byte gets passed into the CommitID constructor as the `commit` arg, which in turn got passed into the old `parseCommit` function which would then throw the error when it saw that the byte wasn't `'0'`.

So now I think the behavior is functionally the same, just potentially with a less descriptive error message for a certain category of malformed CommitID where it uses a single byte for the commit but that byte doesn't contain '0'.  Not a big deal.  If you want to preserve the existing behavior I think you could just change the line this comment is on to say `commit: genesisRemainder[0]`, since that will then get passed into the CommitID constructor, and then down into `parsing.parseCommit`, which will return the same error message that used to be thrown.

Up to you though, I don't think it's a big deal either way.

Reply:
sweet!